### PR TITLE
Porting elementwise metrics to GPU.

### PR DESCRIPTION
**Pull request by trivialfis**

### CODE REVIEW

**Comment by trivialfis** [comment]
TODO:
 - Make sure the parameter configuration is called in every cases.
 - See if it's possible to determine whether to use GPU based on some heuristics.

@RAMitchell I currently don't have any good idea for the second one, but adding another place that can cause troubles for CPU users is not idea either.

**Comment by hcho3** [comment]
The Jenkins webhook is returning Internal Server Error (500). Investigating.

screen shot 2018-11-29 at 8 11 10 pm

**Comment by trivialfis** [comment]
@hcho3 Thanks!

File: src/metric/elementwise_metric.cu [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+
+    thrust::counting_iterator<size_t> begin(0);
+    thrust::counting_iterator<size_t> end = begin + n_data;
+
+    auto s_label = labels.DeviceSpan(device_id);
+    auto s_preds = preds.DeviceSpan(device_id);
+    auto s_weights = weights.DeviceSpan(device_id);
+
+    bool const is_null_weight = weights.Size() == 0;
+    bst_float extra_parameter = policy_.GetExtra();
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/metric/elementwise_metric.cu]: I tried.  There are Metrics classes with a `std::string name_` member, like `EvalError`.

**Comment by RAMitchell** [comment]
Looks good, merge when you are ready.

**Comment by RAMitchell** [comment]
Actually I just remembered one thing: thrust reduce allocates memory, in a multi-GPU setting this will synchronise all GPUs. You will need to pass a temporary allocator to thrust to avoid this.

**Comment by trivialfis** [comment]
@RAMitchell Please help taking a look in the latest commit.  I actually don't understand why it fixes the bug.

**Pull request merged**